DirectEmployers/MyJobs
# Tiny proof of concept for GraphQL.

## What's this PR do?

Allows us to play with GraphQL on some of our data. The endpoint exposed in this branch could conceivably replace a _lot_ of ad hoc code we've written in Reporting and Analytics for running and displaying queries and charts. It's not a good replacement for invoking back end code that does things and changes data. (because there's no good validation story for GraphQL mutations) There is very good in-browser tooling for binding react components to graphql queries.

I'm interested in feedback about this. Obviously don't merge it.

## Where should the reviewer start?

The graphene code in `schema.py`.

I tried a direct-ish connection to the ORM and a datasource.

## How should this be manually tested?

Run `dk rebuilddev`

Go to https://secure.my.jobs/graphql .

If you have some analytics data then this query should be interesting. The pageviews field implements the relay node interface so it gets server side paging for free. I put the relevant fields in the query to help people get started.

```
{
  jobViewRanks {
    browser
    jobViewsExact
    referralApplyRate
    rmPlatformApplyRate
    reachedJobSource
  }
  pageviews(first: 20, after: "YXJyYXljb25uZWN0aW9uOjA=") {
    pageInfo {
      startCursor
      endCursor
      hasPreviousPage
      hasNextPage
    }
    edges {
      node {
        time
        hostname
        path
      }
    }
  }
}
```

## Any background context you want to provide?

I wish I had known about GraphQL when I wrote the new reporting system. It would have saved a lot of time.
